subsys/prefetch/src/main/java/org/commonjava/indy/subsys/prefetch/PrefetchManager.java

```diff
@@ -54,42 +53,83 @@
     @Inject
     private PrefetchConfig config;
 
-    @WeftManaged
-    @ExecutorConfig( named = "prefetch-worker", threads = 5, priority = 1 )
     @Inject
-    private Executor executor;
+    private SpecialPathManager specialPathManager;
 
-    @Inject
-    private Instance<ContentListBuilder> listBuilders;
+    private volatile boolean stopped;
+
+    private final ExecutorService prefetchExecutor =
+            Executors.newFixedThreadPool( 5, new NamedThreadFactory( "Prefetch-Worker", true, 1 ) );
+
+    private final ExecutorService rescanScheduleExecutor =
+            Executors.newSingleThreadExecutor( new NamedThreadFactory( "Prefetch-Rescan-Scheduler", true, 1 ) );
+
+    private final Runnable rescanSchedulingTask = () -> {
+        if ( config.isEnabled() )
+        {
+            while ( !stopped )
+            {
+                if ( !frontier.hasMore() )
+                {
+                    frontier.rescheduleForRescan();
+                    //TODO need to think use a flag to control the triggerWorkers(), and not invoke it in main thread in registerPrefetchStores
+                    triggerWorkers();
```

Comment:
I don't see triggerWorkers() in this PR...I think it's in a previous one, maybe. But in that, do we have a provision for clearing out the downloaded file in the event of a download failure? Or is that using the galley transport api?

Reply:
Currently two another usage of this is in PrefetchWorker.run and PrefetchManager.registerPrefetchStores. But I'm thinking that we should control this for two following purposes:
1. If the triggerWorkers is already running, do not need to run it again
2. Should not run it in main thread because of potential long-running of it could block the thread
The PrefetchWorker.run will trigger many times of this method in parallel, which can cause 1. And the PrefetchManager.registerPrefetchStores is calling in main thread(or a ArtifactStorePostUpdateEvent thread), may cause 2. That's why gave a comment here.